python/triton_mlir/triton_mlir/TritonExtension.cpp

```diff
@@ -354,6 +363,61 @@ createTargetMachine(llvm::Module *module, std::string proc,
   return machine;
 }
 
+/// addPassesToX helper drives creation and initialization of TargetPassConfig.
+static llvm::TargetPassConfig *
+addPassesToGenerateCode(llvm::CodeGenTargetMachineImpl &TM,
+                        llvm::PassManagerBase &PM, bool DisableVerify,
+                        llvm::MachineModuleInfoWrapperPass &MMIWP) {
+  // Targets may override createPassConfig to provide a target-specific
+  // subclass.
+  llvm::TargetPassConfig *PassConfig = TM.createPassConfig(PM);
+  PassConfig->insertPass(&llvm::SIFoldOperandsLegacyID,
+                         &llvm::MaxsMachineFunctionID);
+  // Set PassConfig options provided by TargetMachine.
+  PassConfig->setDisableVerify(DisableVerify);
+  PM.add(PassConfig);
+  PM.add(&MMIWP);
+
+  if (PassConfig->addISelPasses())
+    return nullptr;
+  PassConfig->addMachinePasses();
+
+  PassConfig->setInitialized();
+  return PassConfig;
+}
+
+class CodeGenTargetMachineImplMine : public llvm::CodeGenTargetMachineImpl {
+public:
+  bool addPassesToEmitFileMine(
+      llvm::PassManagerBase &PM, llvm::raw_pwrite_stream &Out,
+      llvm::raw_pwrite_stream *DwoOut, llvm::CodeGenFileType FileType,
+      bool DisableVerify = true,
+      llvm::MachineModuleInfoWrapperPass *MMIWP = nullptr) {
+    // Add common CodeGen passes.
+    if (!MMIWP)
+      MMIWP = new llvm::MachineModuleInfoWrapperPass(this);
+    llvm::TargetPassConfig *PassConfig =
+        addPassesToGenerateCode(*this, PM, DisableVerify, *MMIWP);
+    if (!PassConfig)
+      return true;
+    PassConfig->dump();
+
+    if (llvm::TargetPassConfig::willCompleteCodeGenPipeline()) {
+      if (addAsmPrinter(PM, Out, DwoOut, FileType,
+                        MMIWP->getMMI().getContext()))
+        return true;
+    } else {
+      // MIR printing is redundant with -filetype=null.
+      if (FileType != llvm::CodeGenFileType::Null)
+        PM.add(llvm::createPrintMIRPass(Out));
+    }
+
+    PM.add(llvm::createFreeMachineFunctionPass());
+
+    return false;
+  }
+};
+
```

Comment:
all of this is copy-pasta just to be able to stick `PassConfig->insertPass` above